File: Lab1/Hill Cipher/hill.py
def make_big_boy(text):
    # plain old uppies for hte lowercase
    return ''.join(char.upper() for char in text if char.isalpha())


def hill_cipher_encrypt(plaintext, key_matrix):
    # if odd length, slap on an X
    if len(plaintext) % 2 != 0:
        plaintext += 'X'

    encrypted_text = ""
    for i in range(0, len(plaintext), 2):
        # get the letter values (A=0, B=1, ...)
        ch1 = ord(plaintext[i]) - ord('A')
        ch2 = ord(plaintext[i + 1]) - ord('A')

        # hill cipher math magic
        new_ch1 = (key_matrix[0][0] * ch1 + key_matrix[0][1] * ch2) % 26
        new_ch2 = (key_matrix[1][0] * ch1 + key_matrix[1][1] * ch2) % 26

        # convert back to letters and append to encrypted_text
        encrypted_text += chr(new_ch1 + ord('A'))
        encrypted_text += chr(new_ch2 + ord('A'))

    return encrypted_text


# input plaintext and key matrix
plaintext = input("enter plaintext: ")
key_matrix = []
print("enter 2x2 key matrix (space-separated):")
for _ in range(2):
    key_matrix.append(list(map(int, input().split())))

plaintext = make_big_boy(plaintext)

result = hill_cipher_encrypt(plaintext, key_matrix)
print(f"encrypted text: {result}")

File: Lab1/Hill Cipher/hill.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

void makeBigBoy(char text[]) {
    for (int i = 0; text[i] != '\0'; i++) {
        if (islower(text[i])) {
            text[i] = toupper(text[i]);
        }
    }
}

// hill cipher encryption
void hillCipherEncrypt(char text[], int keyMatrix[2][2]) {
    char encryptedText[100];
    int len = strlen(text);

    // pad with 'X', odd length bhaye
    if (len % 2 != 0) {
        text[len++] = 'X';
        text[len] = '\0';
    }

    // process digraphs
    for (int i = 0; i < len; i += 2) {
        int ch1 = text[i] - 'A';
        int ch2 = text[i + 1] - 'A';

        encryptedText[i] = ((keyMatrix[0][0] * ch1 + keyMatrix[0][1] * ch2) % 26) + 'A';
        encryptedText[i + 1] = ((keyMatrix[1][0] * ch1 + keyMatrix[1][1] * ch2) % 26) + 'A';
    }

    encryptedText[len] = '\0'; // null-terminate
    printf("encrypted text: %s\n", encryptedText);
}

int main() {
    char text[100];
    int keyMatrix[2][2];

    printf("enter plaintext: ");
    fgets(text, sizeof(text), stdin);
    text[strcspn(text, "\n")] = '\0'; // remove newline
    makeBigBoy(text); // make uppercase

    printf("enter 2x2 key matrix (space-separated): ");
    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < 2; j++) {
            scanf("%d", &keyMatrix[i][j]);
        }
    }

    hillCipherEncrypt(text, keyMatrix);

    return 0;
}
